### PR TITLE
feat(bool): add comparison operator to bool

### PR DESCRIPTION
**Pull request by stakach**
Effective works the same way an `Int` comparison would work for `1` and `0` as per the underlying implementation of Bool

### CODE REVIEW

**Comment by asterite** [comment]
Thank you, but I don't think bool and int should be comparable in any sense.

**Comment by docelic** [comment]
> Thank you, but I don't think bool and int should be comparable in any sense.

I believe Stephen meant to say that he added "Bool <=> Bool" comparison in the same way that it exists for "Int <=> Int".

**Comment by asterite** [comment]
Oooh... I see. Yeah, that makes sense.

I'm curious why this isn't there in Ruby. I'm sure there must be a reason, so I'm a bit hesitant about adding this.

**Comment by stakach** [comment]
It works in ruby for me
image

**Comment by asterite** [comment]
Try it with `true <=> false`. You get `nil` instead of `-1` or `1`.

I think in Ruby that means: true has the same order as true, same goes for false against false, but true vs. false are not comparable.

Whether that's okay, I don't know.

**Comment by stakach** [comment]
I just wanted to sort an array with objects that had a bool attribute, I doubt it's all that common just thought it surprising I couldn't with a simple compare

**Comment by asterite** [comment]
I guess this is fine 👍 

I'll wait for others to comment.

**Comment by yxhuvud** [comment]
I don't have any opinion of the feature itself, but it should probably have a spec.

**Comment by bcardiff** [comment]
In ruby you usually end up doing something like `[true, false].sort_by { |x| x ? 1 : 0 }`.
`true <=> true` works  because it is defined in Object using == to identify equality and nothing else AFAIK.
Haskell treat boolean as comparable.

I think we can allow this. I can't come with some undesired behavior. Having `false < true` sound good enough.
We are also making enums comparable.

We should add `include Comparable(self)` to bool.

**Comment by stakach** [comment]
Thanks for the feedback!
`Comparable` and spec added

**Comment by asterite** [comment]
For some reason my body starts itching when I see two boolean values being compared, I don't know why.

I know you can do it in Haskell, and it seems in Java and C# you can use a `compareTo` method on booleans, but for example in C# doing `false < true` doesn't compile. Same for Java. And I think it's the same for every language out there... except Haskell?

But maybe it's fine.

**Comment by asterite** [comment]
Well, apparently you can also compare booleans in Python...

File: src/bool.cr [hunk 1]
```diff
@@ -63,6 +65,18 @@ struct Bool
...
 
+  # Compares this bool against another, according to their underlying value.
+  #
+  # ```
+  # false <=> true  # => -1
+  # true <=> false  # => 1
+  # true <=> true   # => 0
+  # false <=> false # => 0
+  # ```
+  def <=>(other : Bool) : Int32
```
Comment (Fryguy) [inline src/bool.cr:76]: Code-oragnization-wise, can you move this up with the other operators?

File: spec/std/bool_spec.cr [hunk 2]
```diff
@@ -41,6 +41,13 @@ describe "Bool" do
     it { false.to_s.should eq("false") }
   end
 
+  describe "<=>" do
```
Comment (Fryguy) [inline spec/std/bool_spec.cr:44]: Same comment with moving it up near the other operators?

File: src/bool.cr (hunk 1 above)
Comment (stakach) [reply src/bool.cr]: done

File: spec/std/bool_spec.cr (hunk 2 above)
Comment (stakach) [reply spec/std/bool_spec.cr]: done

**Review by jhass — APPROVED** [review]

**Comment by j8r** [comment]
Can it be possible to make `<=>` return `0` if the two values are the same, or `nil` if not?

**Comment by asterite** [comment]
@j8r please no, nil means the two objects are not comparable

File: src/bool.cr [hunk 3]
```diff
@@ -1,10 +1,14 @@
+require "./comparable"
```
Comment (bcardiff) [inline src/bool.cr:1]:
```suggestion
require "comparable"
```

The relative paths are used within each module. AFAIK
Using require "comparable" will use the regular lookup that allows among other things redefining the module via CRYSTAL_PATH. Not a main use case, but is better to not loose that.

**Comment by j8r** [comment]
@asterite are they comparable? We can just say that false is not true. One superior or inferior to the other has little sense in programming.

I don't see a use case for `true < false` for instance.

**Comment by j8r** [comment]
I mean, I don't see any method useful in https://crystal-lang.org/api/master/Comparable.html for Booleans.
Please, prove me wrong :)

**Comment by asterite** [comment]
@j8r https://github.com/crystal-lang/crystal/pull/9863#issuecomment-718753086

**Comment by j8r** [comment]
What's wrong with the @bcardiff solution? And also, why `false` values will be sorted at first, and not `true`? I don't think it is obvious that `false < true`.

**Comment by straight-shoota** [comment]
Implementing `Bool#<=>` was suggested and refused before: #5162
The main reason is that boolean values have no natural ordering. And I think that's a very valid point. I don't see any sense in making a type comparable when the only actual comparability consists of equal values being equal. That's nonsense. You just can't order boolean values.

In some languages it's possible to compare booleans, but that can probably be attributed to the practice of representing boolean values as numbers (for example in C).

**Comment by asterite** [comment]
Yeah, in Haskell Bool is an "enum" type so it naturally follows that there's an order (well, you have to use the `Ord` type class so that's not necessarily true). In Python they are comparable but they aren't numeric values.

I guess the question is: what's the problem if we go forward with this? It's true that bools don't have a natural order. But make a poll and ask people which one they think comes "first", true or false, if you had to guess, for a programming language. My bet is that they will say "false" comes before "true", maybe because they will actually related them to 0 and 1.

**Comment by straight-shoota** [comment]
> My bet is that they will say "false" comes before "true"

Maybe. But that's not an inherent order based on any property of the data type values. It's just an arbitrary order subjectively associated with it. It depends entirely on the use case. So there really shouldn't be a generic implementation just for the sake of it.

> I guess the question is: what's the problem if we go forward with this?

No, the question is: What real world problem does this solve? We shouldn't merge this just because it causes no problem. Only if it solves a specific problem.

**Comment by Sija** [comment]
> Maybe. But that's not an inherent order based on any property of the data type values. It's just an arbitrary order subjectively associated with it. It depends entirely on the use case. So there really shouldn't be a generic implementation just for the sake of it.

@straight-shoota That's true but what's more important here I think is that such order is guaranteed, so if your use case needs sth else you'd just use reversed sort and that's it since there's no 3rd option to choose from anyway...

**Comment by bcardiff** [comment]
As I see this, it's adds a more convenient to handle sorting of Bool types. As such, we can always fallback to `expr ? 1 : 0` instead of `expr` if we don't go forward with this PR.

Regarding the argument of `true < false` or `false < true`. I like the later, but it ease to use `!expr` to sort things the opposite way. Not different as `expr ? 0 : 1` vs `expr ? 1 : 0`. We allow `<=>` on all enums by default based on their int value. I think we should treat booleans the same way here.

I think that by relying on the ternary operator we are forcing developers to discover that trick, when we could avoid them that at no expense.

I think is nicer to deal with nilable elements in array. But is not something that can't be done today.

```crystal
[5, nil, 2].sort_by { |x| {x.nil? ? 0 : 1, x || 0} }
# vs
[5, nil, 2].sort_by { |x| {x.nil?, x || 0} }
```

**Comment by j8r** [comment]
Could Bool in Crystal become an enum, then?
By looking at its methods in https://github.com/crystal-lang/crystal/blob/master/src/bool.cr (like `|` and `&`), it won't be absurd.

**Comment by asterite** [comment]
A question for y'all. What do you expect this to do?

```crystal
[1, 19, 3, 2, 4, 5].sort_by &.even?
```

If you ask me, since I'm sorting by `even?`, I would expect even numbers to come first, then odd numbers to come later. But the result is actually the opposite!

So I think we shouldn't go with this PR forward, it can be confusing.

**Comment by stakach** [comment]
I feel like I would expect odd and even numbers to be grouped together, if the order was undesirable I would use `.odd?`

**Comment by straight-shoota** [comment]
I'm closing this. There are diverse opionions on this topic and no approving consensus.

To summarize the arguments for rejection, I think the most evident point is that lacking any intrinsic order, `true < false` would be as logical as `false > true`. Chosing any order would be arbitrary and lead to surpsises (such as mentioned in https://github.com/crystal-lang/crystal/pull/9863#issuecomment-719644367).
If you want to sort by a boolean value, it should be better to define an explicit order.

**Pull request closed without merging**